gnomad_qc/v4/subset.py

```diff
@@ -1,269 +1,468 @@
-
-    if args.subset_workspaces:
-        terra_workspaces = hl.literal(subset_ht.terra_workspace.lower().collect())
-        subset_ht = meta_ht.filter(
-            terra_workspaces.contains(meta_ht.project_meta.terra_workspace)
-        ).select()
-        logger.info(
-            "Keeping %d samples using the list of terra workspaces.", subset_ht.count()
+        logger.info("Loading the HT with sample IDs to subset...")
+        subset_ht = get_subset_ht(
+            config.subset_samples, config.subset_workspaces, meta_ht
         )
 
-    if args.subset_samples:
-        subset_ht = subset_ht.key_by("s")
-
-    vds = hl.vds.filter_samples(
-        vds, subset_ht, remove_dead_alleles=False if split_multi else True
-    )
-    logger.info(
-        "Final number of samples being kept in the VDS: %d.",
-        vds.variant_data.count_cols(),
-    )
+        logger.info("Checking that all requested sample IDs are in the dataset...")
+        check_subset_ht(subset_ht, vds.variant_data.cols())
 
-    vd = vds.variant_data
-    if split_multi:
-        logger.info("Splitting multi-allelics")
-        vd = vd.annotate_rows(
-            n_unsplit_alleles=hl.len(vd.alleles),
-            mixed_site=(hl.len(vd.alleles) > 2)
-            & hl.any(lambda a: hl.is_indel(vd.alleles[0], a), vd.alleles[1:])
-            & hl.any(lambda a: hl.is_snp(vd.alleles[0], a), vd.alleles[1:]),
+        logger.info("Filtering the VDS to requested samples...")
+        vds = hl.vds.filter_samples(
+            vds, subset_ht, remove_dead_alleles=False if config.split_multi else True
         )
-        vds = hl.vds.split_multi(
-            hl.vds.VariantDataset(vds.reference_data, vd), filter_changed_loci=True
-        )
-    else:
-        logger.info(
-            "Applying min_rep to the variant data MT because remove_dead_alleles in"
-            " hl.vds.filter_samples may result in variants that do not have the minimum"
-            " representation."
-        )
-        vds = hl.vds.VariantDataset(
-            vds.reference_data,
-            vd.key_rows_by(**hl.min_rep(vd.locus, vd.alleles)),
-        )
-
-    if add_variant_qc or pass_only:
-        vd = vds.variant_data
-        ht = release_sites().ht()
-        if pass_only:
-            logger.info("Filtering to variants that passed variant QC.")
-            vd = vd.filter_rows(ht[vd.row_key].filters.length() == 0)
-        if add_variant_qc:
-            logger.info("Adding variant QC annotations.")
-            vd = vd.annotate_rows(
-                **make_variant_qc_annotations_dict(vd.row_key, variant_qc_annotations)
-            )
-        vds = hl.vds.VariantDataset(vds.reference_data, vd)
 
-    if args.vcf or args.dense_mt:
-        logger.info("Densifying VDS")
-        mt = hl.vds.to_dense_mt(vds)
+        if config.output_vcf:
+            logger.info("Densifying VDS for VCF export...")
+            mtds = hl.vds.to_dense_mt(vds)
```

Comment:
I completely forgot about this earlier, but does `vds.to_dense_mt` remove the need for an additional filter after to remove only reference rows? we had to remove these manually in the past with something like
```
mtds = mtds.filter_rows(hl.len(mtds.alleles) > 1)
```

Reply:
Yeah it should remove the ref only itself with this line: https://github.com/hail-is/hail/blob/458d54dbe14a7188720bf706d6acba1114adf77e/hail/python/hail/vds/methods.py#L57